libraries/AP_Scripting/generator/description/bindings.desc

```diff
@@ -124,10 +126,25 @@ singleton AP_Notify alias notify
 singleton AP_Notify method play_tune void string
 singleton AP_Notify method handle_rgb void uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX
 
+include AP_Proximity/AP_Proximity.h
+
+singleton AP_Proximity alias proximity
+singleton AP_Proximity method num_sensors uint8_t
+singleton AP_Proximity method get_object_count uint8_t
+singleton AP_Proximity method get_closest_object boolean float'Null float'Null
+singleton AP_Proximity method get_object_angle_and_distance boolean uint8_t 0 UINT8_MAX float'Null float'Null
```

Comment:
Have you tested this double Null? I think this is the first binding with that format.

Reply:
@IamPete1 Actually I wanted some advice on this. I tested it with REPL, but I couldn't clearly determine if it was working correctly. I'll draft this PR until it's clear.

Reply:
Here is my test for proximity.
Is a pass-by-reference argument a return value in Lua?
![image](https://user-images.githubusercontent.com/22985371/88613919-8fee8980-d0c9-11ea-818b-3ac05c5fe7c6.png)

Reply:
@IamPete1 This is generated code.
```
static int AP_Proximity_get_object_angle_and_distance(lua_State *L) {
    AP_Proximity * ud = AP_Proximity::get_singleton();
    if (ud == nullptr) {
        return luaL_argerror(L, 1, "proximity not supported on this firmware");
    }

    binding_argcheck(L, 2);
    const lua_Integer raw_data_2 = luaL_checkinteger(L, 2);
    luaL_argcheck(L, ((raw_data_2 >= MAX(0, 0)) && (raw_data_2 <= MIN(UINT8_MAX, UINT8_MAX))), 2, "argument out of range");
    const uint8_t data_2 = static_cast<uint8_t>(raw_data_2);
    float data_5003 = {};
    float data_5004 = {};
    const bool data = ud->get_object_angle_and_distance(
            data_2,
            data_5003,
            data_5004);

    if (data) {
        lua_pushnumber(L, data_5003);
        lua_pushnumber(L, data_5004);
    } else {
        lua_pushnil(L);
    }
    return 2;
}

static int AP_Proximity_get_closest_object(lua_State *L) {
    AP_Proximity * ud = AP_Proximity::get_singleton();
    if (ud == nullptr) {
        return luaL_argerror(L, 1, "proximity not supported on this firmware");
    }

    binding_argcheck(L, 1);
    float data_5002 = {};
    float data_5003 = {};
    const bool data = ud->get_closest_object(
            data_5002,
            data_5003);

    if (data) {
        lua_pushnumber(L, data_5002);
        lua_pushnumber(L, data_5003);
    } else {
        lua_pushnil(L);
    }
    return 2;
}
```

Reply:
The generated code looks OK, the case to test would be if it returns false and you try to query the second value. The generator is only pushing a single value to nil. 
```
a, b = proximity:get_object_angle_and_distance()

type(a)
type(b)
a
b
```
Could well be fine as it is

Reply:
![image](https://user-images.githubusercontent.com/22985371/88670632-44b39580-d120-11ea-8380-7e5d5b92c92c.png)

Reply:
![image](https://user-images.githubusercontent.com/22985371/88671218-0074c500-d121-11ea-9ffe-8976c6aec637.png)

Reply:
I don't think you should get a as a number but b as a nil. The should both be numbers or both nil. Looks like a is returning  the input to the function? I think we would just need to add that second lua_pushnil(L) call, that would require a edit to the binding generator.

Reply:
@IamPete1 I'm trying to add sample code. But I found there different return value between REPL and script file.
REPL returns number value for closest_distance but script file return nil.
```
-- This script checks Proximity

function update()
  sensor_count = proximity:num_sensors()
  gcs:send_text(0, string.format("%d proximity sensors found.", sensor_count))

  if sensor_count > 0 then
    object_count = proximity:get_object_count()
    gcs:send_text(0, string.format("%d objects found.", object_count))

    closest_angle, closest_distance = proximity:get_closest_object()
    if closest_distance == nil then
      closest_distance = -1
    end
    gcs:send_text(0, "Closest object at angle "..closest_angle.." distance "..closest_distance)
    
    for i = 0, object_count do
      angle, distance = proximity:get_object_angle_and_distance(i)
      if distance == nil then
        distance = -1
      end
      gcs:send_text(0, "Object"..i.." at angle "..angle.." distance "..distance)
    end
  end

  return update, 2000 -- check again in 0.5Hz
end

return update(), 2000 -- first message may be displayed 1 seconds after start-up
```

![image](https://user-images.githubusercontent.com/22985371/88688074-f60ff680-d133-11ea-934e-1eeca1c55459.png)

![image](https://user-images.githubusercontent.com/22985371/88688124-fd370480-d133-11ea-840b-c61cc0bc6859.png)

Reply:
@IamPete1 I'm trying to add sample code. But I found difference between REPL and script file return value.
When I call get_closest_object function
REPL returns number for closest_distance
![image](https://user-images.githubusercontent.com/22985371/88688818-b39ae980-d134-11ea-8017-e5fadf8fe90b.png)

But script file return nil.
![image](https://user-images.githubusercontent.com/22985371/88688854-beee1500-d134-11ea-88b4-6f31ccf0c9c2.png)

This is my test code.
```
-- This script checks Proximity

function update()
  sensor_count = proximity:num_sensors()
  gcs:send_text(0, string.format("%d proximity sensors found.", sensor_count))

  if sensor_count > 0 then
    object_count = proximity:get_object_count()
    gcs:send_text(0, string.format("%d objects found.", object_count))

    closest_angle, closest_distance = proximity:get_closest_object()
    gcs:send_text(0, "Closest object at angle "..closest_angle.." distance "..closest_distance)
    
    for i = 0, object_count do
      angle, distance = proximity:get_object_angle_and_distance(i)
      gcs:send_text(0, "Object"..i.." at angle "..angle.." distance "..distance)
    end
  end

  return update, 2000 -- check again in 0.5Hz
end

return update(), 2000 -- first message may be displayed 1 seconds after start-up

```

Reply:
@IamPete1 I fixed it manually and got the result below. Is this the result you were saying?
```
static int AP_Proximity_get_object_angle_and_distance(lua_State *L) {
    AP_Proximity * ud = AP_Proximity::get_singleton();
    if (ud == nullptr) {
        return luaL_argerror(L, 1, "proximity not supported on this firmware");
    }

    binding_argcheck(L, 2);
    const lua_Integer raw_data_2 = luaL_checkinteger(L, 2);
    luaL_argcheck(L, ((raw_data_2 >= MAX(0, 0)) && (raw_data_2 <= MIN(UINT8_MAX, UINT8_MAX))), 2, "argument out of range");
    const uint8_t data_2 = static_cast<uint8_t>(raw_data_2);
    float data_5003 = {};
    float data_5004 = {};
    const bool data = ud->get_object_angle_and_distance(
            data_2,
            data_5003,
            data_5004);

    if (data) {
        lua_pushnumber(L, data_5003);
        lua_pushnumber(L, data_5004);
    } else {
        lua_pushnil(L);
        lua_pushnil(L);
    }
    return 2;
}

static int AP_Proximity_get_closest_object(lua_State *L) {
    AP_Proximity * ud = AP_Proximity::get_singleton();
    if (ud == nullptr) {
        return luaL_argerror(L, 1, "proximity not supported on this firmware");
    }

    binding_argcheck(L, 1);
    float data_5002 = {};
    float data_5003 = {};
    const bool data = ud->get_closest_object(
            data_5002,
            data_5003);

    if (data) {
        lua_pushnumber(L, data_5002);
        lua_pushnumber(L, data_5003);
    } else {
        lua_pushnil(L);
        lua_pushnil(L);
    }
    return 2;
}
```

![image](https://user-images.githubusercontent.com/22985371/88691971-7d5f6900-d138-11ea-95d1-bd5409de8a5e.png)

Reply:
Can you try this, https://github.com/tajisoft/ardupilot/pull/1

Reply:
Great!, that is what the change to the binding gen also does, your test script now give the same a REPL?

Reply:
I think the right fix is here: https://github.com/ArduPilot/ardupilot/pull/14932 but I still need to actually get this running and verify.

Reply:
@WickedShell @IamPete1 I tested this PR on #14932 and it worked fine. Thank you for all the help!!